### DISCUSSION TITLE
Add custom linking overrides

### DISCUSSION THREAD

**Discussion by ftphikari** [odin-lang/Odin#1390] [Ideas/Requests]
Currently a number of vendor packages link to libraries be it `system:somelib` or `somelib.lib`. However they are all shared, and require to be distributed with the executable, or to be present in the system. It would be better to just have 1 executable without much dependencies, for which I would use my own library. A couple of solutions I see:

## Solution 1 (not ideal)
Copy the whole package to my project folder, and change every `import` to the library I need.

Problems:
1. In the sdl2 package for example, I would need to change 30+ files, which is doable with `sed -i` but a tad bizzare.
2. If the vendor bindings would change upstream, I would need to manually patch my version of the bindings.
3. Since vendor bindings are already standardized, that would just introduce unneeded redundancy. You might as well have your own `os` package if you don't like the standard one, but that seems like a bit of an overkill for this scenario.

## Solution 2 (proposal)
Introduce a flag/source code tag/etc that lets me change every library link to another, i.e.
```
odin build src -replacelib:system:somelib=myownlib.a;somelib.lib=myownlib.lib
```
or
```
package main

import sdl2 "vendor/sdl2"

#replacelib "system:somelib" "myownlib.a"
#replacelib "somelib.lib" "myownlib.lib"
```

I think adding a flag to the compiler is preferred, since it's unclear how the code would behave if `#replacelib` was called in several places.

**Comment by gingerBill** [comment]
It's not a bad idea to override possible libraries.  I'm not sure on the syntax on whether it should be source level, or command line level, or both.

But in general, the idea should work perfectly fine since it's just a basic text substitution problem.

**Reply by gingerBill** [reply]
`#override_foreign_library "system:somelib" "myownlib.a"` should be fine for a source-level override.

**Reply by Kelimion** [reply]
What about `foreign #config(SDL2_LIB, "sdl2.lib")`?

**Reply by ftphikari** [reply]
Would require you to specify the config and a ~~variable~~ constant everywhere in the source. As i said, that's 30+ files for SDL2. It's also not as flexible as just replacing any library in any place. The package maintainer doesn't have to think about making a constant for it, he just does `system:bloop.lib` and forgets about it - if you want custom version, you do it yourself, and link it yourself.

**Comment by flysand7** [comment]
> It would be better to just have 1 executable without much dependencies, for which I would use my own library

Do you mean replacing dynamic library with a custom-built static library?

**Comment by hgranthorner** [comment]
Bumping this discussion since I ran into this issue recently - another alternative would just be to have the vendored libraries support `#config` for the paths. Using SDL3 as an example:
```odin
// vendor/sdl3/sdl3__foreign.odin
package sdl3

when ODIN_OS == .Windows {
	DEFAULT_LIBRARY_PATH :: "SDL3.lib"
} else {
	DEFAULT_LIBRARY_PATH :: "system:SDL3"
}
LIBRARY_PATH :: #config(SDL3_PATH, DEFAULT_LIBRARY_PATH)

when LIBRARY_PATH != DEFAULT_LIBRARY_PATH {
  // logic for foreign importing and linking with required system libraries, for example:
  when ODIN_OS == .Darwin && LIBRARY_PATH[:len(LIBRARY_PATH) - len(".a")] == ".a" {
    @(export)
    foreign import lib {"system:" + LIBRARY_PATH, "system:Cocoa.framework", "system:OpenGL.framework", "system:IOKit.framework", "system:AVFoundation.framework", "system:CoreAudio.framework", "system:AudioToolbox.framework", "system:CoreHaptics.framework", "system:CoreMedia.framework", "system:Metal.framework", "system:GameController.framework", "system:ForceFeedback.framework", "system:Carbon.framework", "system:CoreVideo.framework", "system:QuartzCore.framework", "system:UniformTypeIdentifiers.framework"}
       // etc.
  }
} else {
  // existing code
}
```
Then users could run the code with `odin run . -define:SDL3_PATH=system:<absolute path to libSDL3.a>`. Obviously the down side is a lot more configuration/maintenance on your end for the various permutations so I can understand not going with this solution, but it wouldn't require any language level changes
